compiler/rustc_codegen_ssa/src/back/linker.rs

```diff
@@ -219,19 +219,44 @@ pub struct GccLinker<'a> {
 }
 
 impl<'a> GccLinker<'a> {
-    /// Argument that must be passed *directly* to the linker
+    /// Passes an argument directly to the linker.
     ///
-    /// These arguments need to be prepended with `-Wl`, when a GCC-style linker is used.
+    /// When the linker is not ld-like such as when using a compiler as a linker, the argument is
+    /// prepended by `-Wl,`.
     fn linker_arg<S>(&mut self, arg: S) -> &mut Self
     where
         S: AsRef<OsStr>,
     {
-        if !self.is_ld {
-            let mut os = OsString::from("-Wl,");
-            os.push(arg.as_ref());
-            self.cmd.arg(os);
+        self.linker_args(iter::once(arg));
+        self
+    }
+
+    /// Passes a series of arguments directly to the linker.
+    ///
+    /// When the linker is ld-like, the arguments are simply appended to the command. When the
+    /// linker is not ld-like such as when using a compiler as a linker, the arguments are joined by
+    /// commas to form an argument that is then prepended with `-Wl`. In this situation, only a
+    /// single argument is appended to the command to ensure that the order of the arguments is
+    /// preserved by the compiler.
+    fn linker_args<I, S>(&mut self, args: I) -> &mut Self
```

Comment:
```suggestion
    fn linker_args(&mut self, args: impl IntoIterator<Item: AsRef<OsStr>>) -> &mut Self
```

Reply:
Or even `&[impl AsRef<OsStr>]`.
`linker_args` is a private function and all the uses are slices, so no need to overengineer.

Reply:
I've also changed this function to use `is_empty` instead of `Peekable` now that the argument is no longer relying on the `IntoIterator` trait.